### ISSUE TITLE
GIMP popups are invisible sometimes

### ISSUE THREAD

**Issue by acoolstraw** [labwc/labwc#317]
GIMP running under xwayland sometimes has the popup windows invisible (for example color change). At first, when you do something that creates a popup like changing text color from the overlay or scaling, the window contents are visible. But sometimes, when you do that thing, it just doesn't render the window contents. I'm not sure why, or how that happens.

labwc version: `labwc 0.5.0-129-g9f181de` (compiled from master, did happen on 0.5.0 too)
wlroots version: `0.15.1-3`

Logs of running `labwc -d 2>log.txt` and reproducing it after a lot of attempts: https://i.troplo.com/i/58708fa34170.txt

![image](https://i.troplo.com/i/84e5632b9973.png )

**Comment by acoolstraw** [comment]
[Clip of it happening, starts at around 0:50](https://i.troplo.com/i/ddcb151dfd62.mp4)

**Comment by Consolatis** [comment]
> [01:14:15] <acoolstraw> this is weird
> [01:14:20] <acoolstraw> bemenu just doesn't appear
> [01:14:33] <acoolstraw> it appears to work - if I run bemenu-run and run a command it does it
> [01:14:39] <acoolstraw> I can launch foot and other apps
> [01:14:47] <acoolstraw> but the menu itself just doesn't render
> [01:15:58] <acoolstraw> it started after I ran Veloren (a Vulkan application, maybe that has something to do with it), and I did something that made the SSD disappear except for the window buttons which stayed there but they were to the left of the dot button (???)
> [01:23:13] <acoolstraw> has anyone had a similar issue with window decorations?

SSD disappearing while still showing the window buttons seems very strange to me.
When toggling decorations it will set the whole SSD tree to disabled and thus the children should *never* be visible because one of their parents got disabled. This almost sounds like some buffers wouldn't get allocated properly which could also explain why an external application like bemenu would suddenly stop rendering as well.

While thinking about that (and also the other issues you experienced).. (x)wayland uses FDs for buffers and the more of them there are it might be possible that the FD limit for labwc and/or xwayland has been reached and thus new buffers can't be allocated properly anymore. #317 (gimp windows) might actually be related to that as well. Setting higher FD limits via ulimit before starting labwc could eventually solve it. #249 might help to solve this automatically if it is indeed the root cause.

AFAIR there was also an issue with systemd restricting the FD limit to some very small number, I remember having to modify something systemwide ages ago so systemd wouldn't restrict usual applications to have insanely small FD limits and thus that might explain why I couldn't reproduce any of the issues (along with me generally not running all the much at once).

**Comment by johanmalm** [comment]
@Consolatis I can reproduce this.

The following patch fixes it, but I'm not sure why. Any thoughts?

```
diff --git a/src/xwayland.c b/src/xwayland.c
index 5662f58..7f51798 100644
--- a/src/xwayland.c
+++ b/src/xwayland.c
@@ -286,7 +286,6 @@ map(struct view *view)
 		return;
 	}
 	view->mapped = true;
-	wlr_scene_node_set_enabled(&view->scene_tree->node, true);
 	if (!view->fullscreen && view->xwayland_surface->fullscreen) {
 		view_set_fullscreen(view, true, NULL);
 	}
@@ -343,6 +342,7 @@ map(struct view *view)
 		      &view->commit);
 	view->commit.notify = handle_commit;
 
+	wlr_scene_node_set_enabled(&view->scene_tree->node, true);
 	view_impl_map(view);
 }
 ```

**Comment by Consolatis** [comment]
hm... my only guess would be that our local commit handler has been assigned at that point whereas before it was not.

Although I'd expect that to not make a difference as we are in a single thread blocking context.
Maybe wlr_scene_node_set_enabled() does some crazy stuff like recursive iterating over commit handlers for all of the tree nodes and if there are none for a surface node break early or something..

**Comment by acoolstraw** [comment]
still happens
![image](https://user-images.githubusercontent.com/45886986/169672656-e451b84c-11a9-4fbe-82ca-5c8f6af3652c.png)

**Comment by johanmalm** [comment]
I've played further with that patch, but the missing window issue persists. Not sure what's causing it.

**Comment by johanmalm** [comment]
This is also fixed for me. Will close.

**Comment by acoolstraw** [comment]
damn, I had it happen to me a day ago. I drew something with the brush, selected the font tool, tried to pick a color there and it was blank.

was able to reproduce it again, although it's not the exact same procedure every time...

![image](https://user-images.githubusercontent.com/45886986/187993731-1bfc629f-8ca9-476e-9934-d5fb52c877e5.png)

**Comment by BitterlyIronic** [comment]
Managed to trigger this with Discord by closing the window and reopening it from the system tray a few times.

![image](https://user-images.githubusercontent.com/44883930/188283423-e7c40468-d1e8-4b97-83b8-6f7fe1c5e8c2.png)

**Comment by Consolatis** [comment]
Maybe this is about some racing condition while \*re\*mapping existing X11 surfaces, GIMP could also keep the unmapped color picker around in case its used later again

**Comment by acoolstraw** [comment]
> Managed to trigger this with Discord by closing the window and reopening it from the system tray a few times.
> 
> ![image](https://user-images.githubusercontent.com/44883930/188283423-e7c40468-d1e8-4b97-83b8-6f7fe1c5e8c2.png)

cute buttons!

**Comment by Consolatis** [comment]
Should be fixed by #526
